src/twisted/web/http_headers.py

```diff
@@ -112,16 +108,39 @@ def __cmp__(self, other):
 
     def _encodeName(self, name: Union[str, bytes]) -> bytes:
         """
-        Encode the name of a header (eg 'Content-Type') to an ISO-8859-1 encoded
-        bytestring if required.
+        Encode the name of a header (eg 'Content-Type') to an ISO-8859-1
+        encoded bytestring if required.  It will be canonicalized and
+        whitespace-sanitized.
 
         @param name: A HTTP header name
 
         @return: C{name}, encoded if required, lowercased
         """
+        if canonicalName := self._canonicalHeaderCache.get(name, None):
+            return canonicalName
+
         if isinstance(name, str):
-            return name.lower().encode("iso-8859-1")
-        return name.lower()
+            bytes_name = name.encode("iso-8859-1")
```

Comment:
I find ternaries easier to read for this sort of thing, and I think that `charmap` better communicates the role of the encoding here, so, if it's all the same to you,

(can't do this as a "suggestion" because of the deleted line")

```python3
bytes_name = name.encode("charmap") if isinstance(name, str) else name
```

Reply:
> Historically, HTTP has allowed field content with text in the
>   ISO-8859-1 charset [[ISO-8859-1](https://datatracker.ietf.org/doc/html/rfc7230#ref-ISO-8859-1)], supporting other charsets only
 >  through use of [[RFC2047](https://datatracker.ietf.org/doc/html/rfc2047)] encoding.  In practice, most HTTP header
 >  field values use only a subset of the US-ASCII charset [[USASCII](https://datatracker.ietf.org/doc/html/rfc7230#ref-USASCII)].
>   Newly defined header fields SHOULD limit their field values to
>   US-ASCII octets.  A recipient SHOULD treat other octets in field
>  content (obs-text) as opaque data.

So I'll stick to iso-8859-1.